### PR TITLE
Improve VirtualConsole performance and correctness

### PR DESCRIPTION
**Pull request by jmcphers**
This change rewrites most of the VirtualConsole class, and its usages, with the goal of solving two underlying issues which have resulted in a variety of bugs:

1. **Insertion into the VirtualConsole is usually O(n^2)**, and is always O(n^2) when the data contains control characters, as the console has no means of pounding over existing data besides a full-on redraw.

2. **Mixing different styles of output in the same console leads to problems**. In the R console, the virtual console is only used for regular output, not error output. For this reason, control characters don't work correctly in error output, and it is not possible for different output classes to co-exist in a single line of output, or overwrite each other.

There are two primary strategies used to resolve these issues:

1. **Let the VirtualConsole connect with the DOM directly**. The VirtualConsole now manages the DOM node which contains all of the textual output. It knows where each character is represented. When data needs to be overwritten, it can locate the DOM nodes which contain that data and update them in place.

2. **Use an efficient data structure to represent the contents of the console.** Rather than maintaining a second buffer which holds the assigned class of every emitted character, a TreeMap is used to remember each region of emitted text. This makes it efficient to locate arbitrary regions of text for replacement when control characters demand it.

### CODE REVIEW

**Comment by jjallaire** [comment]
@jeroenooms

**Comment by kevinushey** [comment]
This is awesome! This will likely also mitigate a large number of issues re: IDE sluggishness that occur when overloading e.g. the console history or the Build pane.

File: src/gwt/src/org/rstudio/core/client/widget/BottomScrollPanel.java [hunk 1]
```diff
@@ -73,7 +80,16 @@ public void onResize()
    @Override
    public void scrollToBottom()
    {
-      DomUtils.scrollToBottom(getElement());
+      scrolling_ = true;
+      try
+      {
+         DomUtils.scrollToBottom(getElement());
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/core/client/widget/BottomScrollPanel.java]: Just out of curiosity, what exceptions were you seeing here?

**Comment by jmcphers** [comment]
@kevinushey It does make things a lot better, but we are still going to need some work in those scenarios. In my testing this at least doubles performance, but appending megabytes of data to the DOM is still very slow and can lock up the browser. The changes to the scroll behavior also help a little, as profiling revealed that once the virtualconsole bottleneck was eliminated, most of the sluggishness was actually caused by `scrollToBottom()` trying to keep up (which is why we now defer that until we've received all the output).

File: src/gwt/src/org/rstudio/core/client/widget/BottomScrollPanel.java (hunk 1 above)
Comment (jmcphers) [inline src/gwt/src/org/rstudio/core/client/widget/BottomScrollPanel.java]: None at all; this is just for safety WRT the guard var `scrolling_`. If that code does throw an exception (and it's JSNI, so hard to make guarantees), `scrolling_` will get stuck at `true` and the panel will stop processing scroll events forever.

File: src/gwt/src/org/rstudio/core/client/VirtualConsole.java [hunk 2]
```diff
@@ -203,99 +126,331 @@ else if (string.length() > maxLength)
...
+         // just append to the existing output stream
+         range.appendRight(text, 0);
       }
       else
       {
-         // consolify just the data to be rendered
-         emitRange(consolify(data), clazz, parent);
+         // create a new output range with this class
+         final ClassRange newRange = new ClassRange(cursor_, clazz, text);
+         parent_.appendChild(newRange.element);
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/core/client/VirtualConsole.java]: Does this need a `null` check? (I notice all other usages of `parent_` are guarded)

Comment (jmcphers) [inline src/gwt/src/org/rstudio/core/client/VirtualConsole.java]: Good question; it does not -- we only call this function when there is a `parent_`. https://github.com/rstudio/rstudio/pull/1035/files#diff-5c5c61a14b365d8fe55a849f020f72ebR318

(some existing `parent_` tests may be superfluous)

This distinguishes the VirtualConsole's "live update" mode (where output is rendered to the DOM as it is received) from the "consolify" case (where a temporary console is constructed, outside the DOM, just to process escape sequences in string).

**Comment by jeroenooms** [comment]
Neat! Looking forward to testing this once it hits the daily builds.

**Comment by jeroenooms** [comment]
Not sure if you already caught this, but there seems to be a strange bug that when I run a command in the terminal that raises an error, the original command is not displayed.

**Comment by jeroenooms** [comment]
I'm sure there is an easy example but I don't have time to debug right now, but this is what I run:

```r
con <- mongolite::mongo("mtcars", url = "mongodb://readwrite:test@ds043942.mongolab.com:43942/jeroen_test")
con$find("some illegal-command")
```

The second line raises an error in C, which seems to wipe out the console history.

**Comment by jmcphers** [comment]
@jeroenooms thanks for reporting! I can't reproduce this on 1.1.93 however. :-/ Is this the same error you're getting?

image

**Comment by jeroenooms** [comment]
For me this happens: https://youtu.be/mVM82uxxzQE

Note how the command that causes the error is not displayed in the console...

**Comment by jmcphers** [comment]
Thanks for making a video; that's helpful! Will look at this in a week or two (am about to be offline for a bit).

**Comment by jmcphers** [comment]
Back online and just pushed a fix for this to master; it will be in tomorrow's daily. Thanks for spotting this!

**Pull request merged**